### PR TITLE
Add `Window::title` getter on Windows and macOS

### PR DESCRIPTION
**Pull request by amrbashir**
- [x] Tested on all platforms changed
	- [x] macOS 
	- [X] Windows
- [X] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [X] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [X] Created or updated an example program if it would help users understand this functionality
- [X] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Porting https://github.com/tauri-apps/tao/commit/c50529b3ee453c73acf36bd7e1cd7c14669951f3

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ And please only add new entries to the top of this list, right below the `# Unre
 
 # Unreleased
 
+- On Windows and macOS, Add `Window::title` getter.
```
Comment (msiglreith) [inline CHANGELOG.md:11]:
```suggestion
- On Windows and macOS, add `Window::title` to query the current window title.
```

File: src/window.rs [hunk 2]
```diff
@@ -952,6 +952,16 @@ impl Window {
     pub fn theme(&self) -> Option<Theme> {
         self.window.theme()
     }
+
+    /// Gets the current title of the window.
+    ///
+    /// ## Platform-specific
+    ///
+    /// - **iOS / Android / x11 / Wayland / Web:** Unsupported.
```
Comment (msiglreith) [inline src/window.rs:960]: should add that we return an empty string in this case.

File: src/platform_impl/windows/window.rs [hunk 3]
```diff
@@ -698,6 +698,14 @@ impl Window {
         Some(self.window_state_lock().current_theme)
     }
 
+    #[inline]
+    pub fn title(&self) -> String {
+        let len = unsafe { GetWindowTextLengthW(self.window.0) };
+        let mut buf = vec![0; len as usize];
```
Comment (msiglreith) [inline src/platform_impl/windows/window.rs:704]: don't we need to account for the `\0` terminator here? (haven't tested it so far)

Comment (amrbashir) [reply src/platform_impl/windows/window.rs:704]: yeah you're correct

File: src/window.rs (hunk 2 above)
Comment (amrbashir) [reply src/window.rs:960]: Sure

File: src/platform_impl/windows/window.rs [hunk 4]
```diff
@@ -698,6 +698,14 @@ impl Window {
         Some(self.window_state_lock().current_theme)
     }
 
+    #[inline]
+    pub fn title(&self) -> String {
+        let len = unsafe { GetWindowTextLengthW(self.window.0) };
+        let mut buf = vec![0; (len + 1) as usize];
+        unsafe { GetWindowTextW(self.window.0, buf.as_mut_ptr(), len) };
+        String::from_utf16_lossy(&buf[..(len - 1) as _])
```
Comment (kchibisov) [inline src/platform_impl/windows/window.rs:706]: Isn't what you're doing is basically `OsString` here? Though, you might be fine with raw buffers.

Comment (amrbashir) [reply src/platform_impl/windows/window.rs:706]: do you want to the API to return `OsString` instead or you want me to go through `OsString` to `String`

Comment (kchibisov) [reply src/platform_impl/windows/window.rs:706]: OsString to String.

Comment (amrbashir) [reply src/platform_impl/windows/window.rs:706]: Wouldn't that result in extra allocation?

Comment (kchibisov) [reply src/platform_impl/windows/window.rs:706]: I guess it's fine then.

File: src/platform_impl/windows/window.rs [hunk 5]
```diff
@@ -698,6 +698,14 @@ impl Window {
         Some(self.window_state_lock().current_theme)
     }
 
+    #[inline]
+    pub fn title(&self) -> String {
+        let len = unsafe { GetWindowTextLengthW(self.window.0) };
+        let mut buf = vec![0; (len + 1) as usize];
+        unsafe { GetWindowTextW(self.window.0, buf.as_mut_ptr(), len) };
+        String::from_utf16_lossy(&buf[..len as _])
```
Comment (msiglreith) [inline src/platform_impl/windows/window.rs:706]: Let's use the `util::decode_wide` function.
I tested it on the `window` example and it dropped the `!` at the end and includes a `\0`.
Below an implementation suggestion which returns the expected string.
```rust
let len = unsafe { GetWindowTextLengthW(self.window.0) } + 1; // incl. NULL terminator
let mut buf = vec![0; len as usize];
unsafe { GetWindowTextW(self.window.0, buf.as_mut_ptr(), len) };
util::decode_wide(&buf).to_string_lossy().to_string()
```

File: src/window.rs [hunk 6]
```diff
@@ -952,6 +952,16 @@ impl Window {
     pub fn theme(&self) -> Option<Theme> {
         self.window.theme()
     }
+
+    /// Gets the current title of the window.
+    ///
+    /// ## Platform-specific
+    ///
+    /// - **iOS / Android / x11 / Wayland / Web:** Unsupported. Always return an empty string.
```
Comment (msiglreith) [inline src/window.rs:960]:
```suggestion
    /// - **iOS / Android / x11 / Wayland / Web:** Unsupported. Always returns an empty string.
```

File: src/platform_impl/windows/window.rs (hunk 5 above)
Comment (amrbashir) [reply src/platform_impl/windows/window.rs:706]: Doesn't that result in an extra allocation?

Comment (msiglreith) [reply src/platform_impl/windows/window.rs:706]: fine for me - it's used in other places already and this function shouldn't be called a lot imo

Comment (amrbashir) [reply src/platform_impl/windows/window.rs]: done

**Review by madsmtm — APPROVED** [review]
macOS impl looks good, and I tested it and it works fine, thanks!

**Review by msiglreith — APPROVED** [review]

**Review by kchibisov — APPROVED** [review]

**Pull request merged**